### PR TITLE
feat(io): JSON I/O processing for Claude Code hooks (Issue #55)

### LINKED ISSUE #55: 🎯 MAOS CLI: JSON Input/Output Processing
## 🎯 MAOS CLI: JSON Input/Output Processing

Part of PRD-02 implementation. This issue implements high-performance JSON processing for stdin/stdout communication with Claude Code hooks.

### 📋 Scope

This issue covers JSON handling from PRD-02 lines 139-215:

- [ ] **Stdin Processing Infrastructure**
  - [ ] `StdinProcessor` struct with buffer management
  - [ ] Async stdin reading with timeout support
  - [ ] Buffer reuse for multiple operations
  - [ ] Configurable max input size limits

- [ ] **JSON Parsing Layer**
  - [ ] High-performance JSON deserialization
  - [ ] Support for all hook message types
  - [ ] Error handling for malformed JSON
  - [ ] Streaming support for large inputs

- [ ] **Hook Message Types**
  - [ ] `HookMessage` enum with all variants
  - [ ] `PreToolUse` message structure
  - [ ] `PostToolUse` message structure  
  - [ ] `Notification` message structure
  - [ ] `Stop` message structure
  - [ ] `UserPromptSubmit` message structure

- [ ] **Output Formatting**
  - [ ] JSON response serialization
  - [ ] Error message formatting
  - [ ] Optional pretty-printing for debugging

#### Stdin Processor
```rust
use tokio::io::{AsyncReadExt, stdin};
use std::time::Duration;

pub struct StdinProcessor {
    buffer: Vec<u8>,
    max_size: usize,
}

impl StdinProcessor {
    pub fn new(max_size: usize) -> Self {
        Self {
            buffer: Vec::with_capacity(8192),
            max_size,
        }
    }
    
    pub async fn read_json<T>(&mut self) -> MaosResult<T>
    where
        T: DeserializeOwned,
    {
        // Read with 100ms timeout
        let input = tokio::time::timeout(
            Duration::from_millis(100),
            self.read_to_buffer()
        ).await.map_err(|_| MaosError::Timeout {
            operation: "stdin_read".to_string(),
            timeout_ms: 100,
        })??;
        
        // Parse JSON
        serde_json::from_slice(&input)
            .map_err(|e| MaosError::Json(e))
    }
}
```

#### Hook Message Types
```rust
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "hook_event_name")]
pub enum HookMessage {
    #[serde(rename = "pre_tool_use")]
    PreToolUse {
        session_id: String,
        transcript_path: PathBuf,
        cwd: PathBuf,
        tool_name: String,
        tool_input: serde_json::Value,
    },
    
    #[serde(rename = "post_tool_use")]
    PostToolUse {
        session_id: String,
        transcript_path: PathBuf,
        cwd: PathBuf,
        tool_name: String,
        tool_result: serde_json::Value,
    },
    
    #[serde(rename = "notification")]
    Notification {
        session_id: String,
        transcript_path: PathBuf,
        cwd: PathBuf,
        message: String,
    },
    
    // ... other variants
}
```

### 📦 Dependencies

**Required crates:**
- `serde` (with derive feature)
- `serde_json`
- `tokio` (with io-util feature)
- `bytes` (for efficient buffer management)

**Blocks:**
- Issue #56: Command Dispatcher (needs message types)
- Issue #57: Exit Code Management (needs error types)

**Depends on:**
- Issue #54: CLI Command Structure (completed)
- PRD-01: Error types and MaosResult

### ✅ Success Criteria

1. **Parsing Performance**
   - Small messages (1KB): <100μs
   - Medium messages (10KB): <500μs  
   - Large messages (100KB): <2ms

2. **Memory Efficiency**
   - Buffer reuse across operations
   - No unnecessary allocations
   - Streaming for large inputs

3. **Error Handling**
   - Graceful handling of malformed JSON
   - Clear error messages
   - Timeout protection

4. **Testing**
   - Unit tests for all message types
   - Round-trip serialization tests
   - Performance benchmarks
   - Edge case handling

### 🚀 Performance Targets

- **JSON Parse Time**: <1ms for typical messages (1-10KB)
- **Memory Usage**: <500KB for buffers
- **Timeout**: 100ms max wait for stdin

### 📚 References

- PRD-02: `docs/prd/PRD-02-cli-structure.md` (lines 139-215)
- PRD-01: Hook message specifications
- serde_json: https://docs.rs/serde_json/
- tokio AsyncRead: https://docs.rs/tokio/

### 💡 Implementation Notes

1. Start with basic stdin reading
2. Add timeout protection
3. Implement message type definitions
4. Add serialization/deserialization
5. Optimize with buffer reuse
6. Add comprehensive tests
7. Benchmark performance

Consider using `simd-json` for additional performance if needed!

### PR DESCRIPTION
**Pull request by clafollett**
## Summary

Implements high-performance JSON input/output processing for all 8 Claude Code hook types with exceptional performance and comprehensive security validation.

## 🚀 Performance Results

- **Small (1KB)**: 316ns (316x better than 100μs target)
- **Medium (10KB)**: 1.35μs (370x better than 500μs target)  
- **Large (100KB)**: 9.1μs (220x better than 2ms target)

## 🛡️ Security Features

- JSON depth validation prevents JSON bombs
- Configurable size limits (default 10MB)
- Timeout protection (5s processing, 100ms per read)
- Input validation for all hook types

### Core Components
- **StdinProcessor**: Async JSON processing with buffer reuse and security hardening
- **HookInput**: Type-safe Claude Code message format with validation
- **HookConfig**: Security & performance configuration integration
- **Comprehensive testing**: 33 unit tests + integration tests + benchmarks

### Supported Hook Types
- `pre_tool_use`, `post_tool_use` (tool lifecycle)  
- `notification`, `user_prompt_submit` (user interaction)
- `stop`, `subagent_stop` (agent lifecycle)
- `pre_compact`, `session_start` (session management)

### TDD Methodology
- **RED**: Failing tests for all 8 hook types and performance benchmarks
- **GREEN**: Implementation achieving 200-300x performance targets  
- **REFACTOR**: Config integration & security hardening

## 📚 Documentation

All APIs are comprehensively documented with examples and usage patterns.

## 🧪 Test Coverage

- **79 unit tests**: All passing
- **33 IO-specific tests**: Complete hook type coverage
- **4 integration tests**: End-to-end validation  
- **Benchmark tests**: Performance validation
- **318 total tests**: All passing

## Test plan

- [x] All 8 Claude Code hook types parse correctly
- [x] Security validation prevents JSON bombs and oversized inputs
- [x] Performance targets exceeded by 200-300x
- [x] Config integration works seamlessly
- [x] All tests pass including pre-commit hooks
- [x] Complete API documentation with examples

Resolves #55

🤖 Generated with [Claude Code](https://claude.ai/code)

### CODE REVIEW

File: docs/reviews/code-review-json-io-processing-2025-08-15.md [hunk 1]
```diff
@@ -0,0 +1,490 @@
+# JSON I/O Processing Security & Architecture Review
+**Issue #55: High-Performance JSON Processing for Claude Code Hooks**
+
+**Reviewer**: Code Review Agent  
+**Date**: 2025-08-15  
+**Review Scope**: Comprehensive security and architecture review  
+**Status**: ⚠️ **Major Issues Found - Do Not Merge Without Fixes**
```
[the remark this answers is not in the document]
Comment (clafollett) [reply docs/reviews/code-review-json-io-processing-2025-08-15.md]: None issue. This was a local code review which was already accounted for. I have removed these documents so as not to confuse Copilot.

File: docs/reviews/code-review-json-io-processing-2025-08-15.md [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+
+## Executive Summary
+
+The JSON I/O processing implementation shows solid engineering fundamentals but contains **critical security vulnerabilities** and **performance gaps** that must be addressed before merging. While the architecture is well-designed and test coverage is comprehensive, several high-severity issues require immediate attention.
+
+### 🚨 Critical Issues (MUST FIX)
+1. **JSON Bomb DoS Vulnerability**: No protection against deeply nested JSON structures
```
[the remark this answers is not in the document]
Comment (clafollett) [reply docs/reviews/code-review-json-io-processing-2025-08-15.md]: None issue. This was a local code review which was already accounted for. I have removed these documents so as not to confuse Copilot.

File: crates/maos/src/io/messages.rs [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+                if let Some(trigger) = &self.trigger
+                    && trigger != "manual"
+                    && trigger != "auto"
+                {
+                    return Err(MaosError::InvalidInput {
+                        message: format!(
+                            "Invalid trigger value: {}. Must be 'manual' or 'auto'",
+                            trigger
+                        ),
+                    });
```
[the remark this answers is not in the document]
Comment (clafollett) [reply crates/maos/src/io/messages.rs:160]: This is legit code. We are on edition 2024 which has been stable for a while now and the `let` assignment feature has been stable for 2 years. I'm not fixing this.
